trace-utils/src/send_data.rs

```diff
@@ -151,79 +302,508 @@ impl SendData {
             req = req.header(*key, value);
         }
 
-        async fn send_request(
-            req: HttpRequestBuilder,
-            payload: Vec<u8>,
-        ) -> Result<Response<Body>, SendRequestError> {
-            let req = req
-                .body(Body::from(payload))
-                .map_err(|e| SendRequestError::Any(anyhow!(e)))?;
-
-            Client::builder()
-                .build(connector::Connector::default())
-                .request(req)
-                .await
-                .map_err(SendRequestError::Hyper)
-        }
+        req
+    }
+
+    async fn send_request(
+        &self,
+        req: HttpRequestBuilder,
+        payload: Vec<u8>,
+    ) -> Result<Response<Body>, SendRequestError> {
+        let req = req
+            .body(Body::from(payload))
+            .map_err(|e| SendRequestError::Any(anyhow!(e)))?;
 
+        Client::builder()
+            .build(connector::Connector::default())
+            .request(req)
+            .await
+            .map_err(SendRequestError::Hyper)
+    }
+
+    async fn send_with_protobuf(&self) -> SendDataResult {
         let mut result = SendDataResult::new();
 
-        if target.api_key.is_some() {
-            req = req.header("Content-type", "application/x-protobuf");
+        let agent_payload = construct_agent_payload(self.tracer_payloads.clone());
```

Comment:
I can't figure out a way to avoid this `clone` 😢 The _raw_ mapping of `proto` definitions to `rust` structs make this messy.